### ISSUE TITLE
No audio on WiiU unless a filtering mode is set

### ISSUE THREAD

**Issue by fr500** [LIJI32/SameBoy#29]
Users report that for some reason they get no audio on WiiU unless they set the High Pass Filter to something other than off.

Maybe an endian-ness issue?

**Comment by LIJI32** [comment]
Could be an endianness issue considering I've only ever tested SameBoy on x86 (32/64) and ARM64, but it doesn't explain why the High Pass Filter changes anything. Sadly I'm not familiar with the Wii U Homebrew scene so I'll be working a bit blindly here, but if that won't work I do have a Wii U I can probably use for development and testing, although it will take some more time.

**Comment by fr500** [comment]
Other than that is almost fullspeed in GB, I'm gonna set the resampler to 44100 on WiiU to see if I can get better results

**Comment by LIJI32** [comment]
For 0.11 I do plan redoing a lot of things that make SameBoy slower than other emulators, so hopefully these kinds of optimizations won't be required.

**Comment by LIJI32** [comment]
I think I have an idea why audio is broken and it's probably a more direct cause – I guess the Wii U's PCM-range is lower than what is expected from a standard 16-bit PCM audio (Or it might gets amplified and clipped). Because the Game Boy's audio has a DC offset before the high pass filter is applied (by the hardware itself or the emulator), any clipping will reduce the audible volume, and extreme clipping will effectively mute it.

Basically I can find out what's the actual max amplitude is and fix it in SameBoy for the Wii U build, but if this is in fact the issue it might be wiser to fix it in RetroArch itself because the clipping might affect other cores as well.

**Comment by LIJI32** [comment]
I've looked at the RetroArch's Wii U audio driver and I didn't find anything obvious that would amplify the audio and cause any clipping, and not having a Wii U set up for development isn't helping. Can you see if redefining `MAX_CH_AMP` (in `apu.h`) to a lower value (`0x1FE0 / 2`? `0x1FE0 / 4`?) solves anything? If it does, it means either RetroArch or the Wii U system software (OS or libraries) amplify cores audio output with clipping, and while it mostly affects SameBoy with High-pass filter disabled, it can also affect other cores and should probably be fixed in RetroArch. If that proves difficult it can be fixed specifically for SameBoy by redefining `MAX_CH_AMP` to a lower value when compiling for the Wii U.

**Comment by fr500** [comment]
Sure thing, I'll let you know

**Comment by fr500** [comment]
You're right, there's audio now

**Comment by LIJI32** [comment]
Great news, so now we know what's causing the problem. I'll go open an issue for RetroArch so it's fixed for all cores.

**Comment by LIJI32** [comment]
I close this issue as it's temporarily fixed by an #ifdef. I added a comment to remove that hack once libretro/Retroarch#6252 is fixed.
